docs/src/references/synchronize-daemons.md

```diff
@@ -0,0 +1,116 @@
+and solutions to avoid this synchronization by relying on dynamic runtime health status.
+
+# Problem statement
+
+Thin-edge.io has three main issues for synchronizing the daemons
+
+1. The thin-edge daemons depend on another daemon's liveness to delegate or request some work.
+ For example, tedge-mapper-c8y depends on tedge-agent. When the tedge-mapper-c8y is started,
+ the mapper sends a request to get the existing software list on `tedge/commands/req/software/list` topic,
+ if the tedge-agent is not up and running then the request message will be lost. So, it's better to check if the agent
+ is up and running before sending the request to get the software list.
+
+2. The thin-edge daemons are dependent on the liveness of the bridge to communicate with the cloud.
+ For example, when the tedge-mapper-c8y comes up, it will try to get the jwt token. The request will be sent to c8y cloud
+ over c8y/s/dat, if the bridge is not up then the request message will be lost.
+ So, better to check the status of the bridge before sending the request to the c8y cloud.
+
+3. Some of the daemons are dependent on the `tedge-mapper-c8y`(cloud mapper) file system to create the supported operations file.
+
+# Proposed solutions and their Pros and Cons
+
+## Using the `health check` mechanism to synchronize `tedge-mapper-c8y` and `tedge-agent`
+
+- The `tedge-mapper-c8y` can use the health-check mechanism to detect the health of the `tedge-agent` and then send the request to the `tedge-agent`.
+- When the mapper starts it will subscribe to the `tedge/health/tedge-agent` topic.
+- When the `tedge-agent` starts, it will publish a `health` status (up/down) message on the `tedge/health/tedge-agent` topic.
+- Based on this `tedge-agent` health status message, the mapper can create the supported operations in the `/etc/tedge/operations/c8y` directory.
+- Check if the `c8y-bridge` is up before sending the init messages, including `supported operations` list.
+- Mapper will send the supported operations list to the c8y cloud.
+- Mapper will send the request `software list` to `tedge-agent`.
+- Once it receives a software list from `tedge-agent`, the mapper will send the list to the c8y cloud.
+
+### Communicating with Cumulocity over HTTP
+
+The `tedge-mapper-c8y` needs `internal-id` of the `thin-edge` device from Cumulocity to communicate over HTTP.
+When the tedge-mapper-c8y starts, it will request the `internal id` of the `thin-edge` device from cumulocity cloud.
+Which will be used by the mapper for further communication with the c8y cloud over HTTP.
+
+To get the `internal id` over HTTP, the mapper needs the `JWT` token, which was retrieved over the `MQTT`.
+So, before sending the request to retrieve the `JWT` token on the `c8y/s/uat` topic,
+the mapper must check the health of the `c8y bridge`.
+When the bridge is up, it can send a request for the JWT token.
+Then the mapper will receive the `JWT` token on the `c8y/s/dat` topic.
```

Comment:
> So, before sending the request to retrieve the `JWT` token on the `c8y/s/uat` topic,
the mapper must check the health of the `c8y bridge`.

This writing gives a hint of the problem but is quite far of how things can work. The mapper *reacts* on events telling the bridge is up and takes these events as an opportunity to fetch the internal device id if not already done.